test/e2e/basic/ping-pong.go

```diff
@@ -2,160 +2,168 @@ package e2e
 		gomega.Expect(err).NotTo(gomega.HaveOccurred())
 
 		for _, pod := range pods.Items {
-			var log string
-			//run command in pod
-			ginkgo.By("Ensuring that ping-pong is working")
-			if pod.Status.Phase == v1.PodRunning {
-				log, err = GetPodLogs(pod.Namespace, pod.Name, "", false)
-				gomega.Expect(err).NotTo(gomega.HaveOccurred())
-				if !strings.Contains(log, "Ping") || !strings.Contains(log, "Pong") {
-					warn.OnError(errors.New("Did not see Ping or either Pong, waiting"))
-				}
-
-				if strings.Contains(log, "Ping") && strings.Contains(log, "Pong") {
-					ginkgo.By("Found Ping, Pong in logs, done")
-					return
-				}
+			log, err := getPodLogs(clientSet, pod.Namespace, pod.Name)
+			if err != nil {
+				return err
+			}
+			if !strings.Contains(log, "Ping") || !strings.Contains(log, "Pong") {
+				continue
+			}
+
+			if strings.Contains(log, "Ping") && strings.Contains(log, "Pong") {
+				ginkgo.By("Found Ping, Pong in logs, done")
+				return nil
 			}
 		}
 	}
-
 }
 
-func specialResourceDelete(cs *framework.ClientSet, cl client.Client, path string) {
-	SRName := "ping-pong"
-	ginkgo.By("deleting " + SRName)
-	sr, err := ioutil.ReadFile(path)
+func pingPongDelete(cs *framework.ClientSet, cl client.Client) error {
+	sr, err := ioutil.ReadFile(pingPongChartPath)
 	if err != nil {
-		panic(err)
+		return err
 	}
-	framework.DeleteFromYAMLWithCR(sr, cl)
+	return framework.DeleteFromYAMLWithCR(sr, cl)
+}
 
-	ginkgo.By(fmt.Sprintf("Confirming %s is deleted", SRName))
-	err = wait.PollImmediate(pollInterval, waitDuration, func() (bool, error) {
+func waitPingPongDeleted(cl client.Client) error {
+	err := wait.PollImmediate(pollInterval, waitDuration, func() (bool, error) {
 		specialresources := &srov1beta1.SpecialResourceList{}
-		err = cl.List(context.TODO(), specialresources, []client.ListOption{}...)
+		err := cl.List(context.Background(), specialresources, []client.ListOption{}...)
 		if err != nil {
-			return false, fmt.Errorf("couldn't get SpecialResourceList: %v", err)
+			return false, err
 		}
-
 		for _, n := range specialresources.Items {
-			if n.ObjectMeta.Name == SRName {
+			if n.ObjectMeta.Name == pingPongName {
 				return false, nil
 			}
 		}
-
 		return true, nil
 
 	})
-	if err != nil {
-		explain = err.Error()
-	}
-	gomega.Expect(err).NotTo(gomega.HaveOccurred(), explain)
+	return err
 }
 
-func specialResourceCreate(cs *framework.ClientSet, cl client.Client, path string) {
-
-	ginkgo.By("creating ping-pong")
-	sr, err := ioutil.ReadFile(path)
+func waitPingPongReady(clientSet *kubernetes.Clientset) error {
+	watch, err := clientSet.CoreV1().Pods(pingPongName).Watch(context.Background(), metav1.ListOptions{})
 	if err != nil {
-		panic(err)
+		return err
 	}
-	framework.CreateFromYAML(sr, cl)
+	quitChan := make(chan struct{})
+	go func() {
+		numPodsRunning := 0
+		for event := range watch.ResultChan() {
+			p, _ := event.Object.(*v1.Pod)
```

Comment:
I think `_` is a boolean indicating the outcome of the type conversion here. Should we check it?

Reply:
Forgot this one.